### DISCUSSION TITLE
Location of test_codes/real_time_clock/code.py?

### DISCUSSION THREAD

**Discussion by mchopping** [STELLA-Landsat/STELLA#30] [Debugging]
My first STELLA-Q is almost ready for use! It is returning messages in the Mu serial panel about the status of the unit, with several lines per cycle. However, I can't yet set the time or test the RTC. In "Step 2: Setting the Time" in the STELLA-Q programming instructions (Apple Setup) it has:

"6. In another Finder/Explorer window your computer, in the code files that you downloaded for programming the STELLA, open the folder called test_codes. In there you should find a folder called real_time_clock, and in that folder you should find a file named code.py"

But 'STELLA-Q_code_and_libraries.zip' only has lib/ and code.py. I assume this code.py file has different contents to test_codes/real_time_clock/code.py, since the timestamp = time.struct_time...  is **above** the if False statement, not a few lines below it (see steps 11 - 13).  I tried editing the code anyway and saving the file at the requisite time, but to no avail (the year comes out as 2145). 

It looks like I need to locate the time set/test version of code.py to make this work. Does anyone know where I can find test_codes/real_time_clock/code.py so I can set the time?

**Comment by mchopping** [comment]
Success! The [test_codes/real_time_clock/] code.py was located in the S2-Software/S2_test_codes/04. real_time_clock/  I copied this to the CIRCUITPY drive after running the STELLA-Q_code_and_libraries/code.py to get the STELLA started, then loaded/edited the RTC test code.py and was able to set the date and time and see them reported in the Mu Serial pane.

**Comment by paulmirel** [comment]
Great! Thanks for the follow up!

Apologies for the incomplete section on STELLA-Q.

**Reply by Landsat-Mike** [reply]
packaged the software with the STELLA-Q test codes and popped them up there

**Reply by mchopping** [reply]
Hi Mike, Paul, We're having issues with the STELLA-Q programming. In the new (https://landsat.gsfc.nasa.gov/wp-content/uploads/2023/03/STELLA-Q-part-3-Operating-instructions.txt) operating instructions, it appears that the time and date can be set at run time. However, this does not work with the updated `code.py` file from the August 31 zip for the STELLA-Q code. Using this new code, the STELLA-Q never boots. Looking through this new `code.py` file, I see that the libraries for the 726x sensors have been replaced with a library for 7265 sensors (which is not a part of the STELLA-Q!). Editing out this import and importing the old 726x libraries allows the STELLA-Q to boot and operate, but we are still unable to set the date/time at runtime (it would be really nice to be able to do this!). We have tried using all combinations of the old/new (8.1.0/8.2.0) binaries with the old/new `code.py` and `lib`. Any usage of the new `code.py` would fail. Please advise! --Mark

**Reply by Landsat-Mike** [reply]
Hey Mark, sorry for the inconvenience it is possible I uploaded the code for the Q2 and not the Q. I will upload the correct code and things should hopefully be correct.  - Mike

**Reply by paulmirel** [reply]
Hi Mark, 

I reviewed the code, repaired the library calls, and updated the version number to 2.1.3. Mike is going to upload that version to the STELLA-Q part of the repository. 

P

**Reply by mchopping** [reply]
Hi Mike, hi Paul, Thank you! We're really looking forward to the students building and using these kits very soon (all the parts ordered have now come in!).

**Comment by mchopping** [comment]
Thanks!

Best wishes,

 Mark

**Comment by mchopping** [comment]
Hi Paul, Mike,

Thanks!  We were scratching our heads. I wondered if there is
a compatibility table available to show the firmware / CircuitPy OS / test
and operating codes for the various versions. For example, does the
STELLA-Q I built in the summer and time-stamped using the test code
have the same operating instructions, apart from setting the date and time?

Best wishes,

 Mark

**Comment by paulmirel** [comment]
Hi Mark, 

I intend that the STELLA-Qs all work the same, running the latest device software. They should work regardless of the CircuitPython operating system, so long as it's in the 8x, or at least 7x version. The library structure changed from 6x to 7x, so a 6x version of CircuitPython likely won't work. 

I try to keep them all up to date, so I'm not tracking versions on a compatibility table. The operating instructions should be the same, but I may have missed some changes from version to version. 

I recommend, if the latest version is working properly for you, that you upgrade them all. If it isn't, please let me know here and I'll go about repairing the code.

**Comment by mchopping** [comment]
Hi Paul,

Thanks, received and understood. I will test and if all is OK, update the
July instrument with the current setup. I'm hoping to distribute the kits
to students next Monday.

Best wishes,

 Mark

**Reply by mchopping** [reply]
10/8/23: Just updating here. We got the new STELLA-Q code working! We had a couple of issues with a corrupted SD card and the early October code. The import for the set_date_time function from stella_q_utilities was commented out on line 48 in code.py. Also, the logic for setting the time and date was commented out on line 424 and replaced with 'if False:', seemingly to disable this functionality. We removed the comments, therefore importing the set_date_time function and replaced the 'if FALSE' on line 424 with the commented out logic, which has allowed us to access the nice "hold down the orange button and set the date and time" functionality. Progress (probably, unless this breaks something else; we also noted a Feather Express memory constraint that we bumped into occasionally).

**Comment by paulmirel** [comment]
Hi Mark, 

Thanks for your persistence, and many apologies that you needed to go into the code at all. I'll do a thorough review tomorrow, to try to get everything both correct and synchronized. 

P

**Comment by mchopping** [comment]
Hi Paul, Thanks (no worries, it's character building!). Best wishes --Mark

**Reply by mchopping** [reply]
10/10/23 update: Last night, the students here built out the circuitry for STELLA-Qs `#1` - `#9` , but they would not boot. Our STELLA-Q `#10` was the one we had working after fixing the date/time issue; we assumed the updated code.py would be OK on the others. The only difference between the code.py on `#10` and the updated code you sent last week was that we had added some lines to collect and show free memory:

> [Marks-MacBook-Pro:~/....../STELLA-Q_FILES] chopping% diff /Users/chopping/Downloads/SQ_SELF_EDIT_GOOD/gooder\ codepy/code.py /Users/chopping/Downloads/SQ_SELF_EDIT_GOOD/code.py 
> 48,50d47
> < mfree = gc.mem_free()
> < print("start memory free {:.1f} kB".format( mfree/1000 ))

I brought `#9` home and added those lines into the code.py and it boots just fine now. I've cycled it on/off a bunch of times without issue, so I think all I need to do is apply that change to the others.

**Comment by paulmirel** [comment]
I went in other directions today, didn't get to the SQ review. Planning to do so tomorrow. 

I've noticed that sometimes they won't boot while connected to the host computer. They may boot if you turn them off, and then turn them on while unplugged from the computer. 

Nice work with your strategy on coping with the memory limits!

**Comment by paulmirel** [comment]
[STELLA-Q_v2_1_4.zip](https://github.com/STELLA-Landsat/STELLA/files/12872936/STELLA-Q_v2_1_4.zip)

I cleaned up the code, found that I had enough available RAM, and restored date and time set on orange hold.
